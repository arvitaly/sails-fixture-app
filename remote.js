"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = require("child_process");
function start(config = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const app = new RemoteApp(config);
        yield app.start();
        return app;
    });
}
exports.start = start;
function lift(config = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const app = new RemoteApp(config);
        yield app.lift();
        return app;
    });
}
exports.lift = lift;
class RemoteApp {
    constructor(config = {}) {
        this.config = config;
        this.commands = {};
        this.id = 0;
        this.child = child_process_1.fork(__dirname + "/remote-client");
        this.child.on("message", (data) => {
            switch (data.type) {
                case "log":
                    console.warn(data.data);
                    break;
                case "resolve":
                    this.commands[data.id].resolve(data.data);
                    break;
                case "reject":
                    this.commands[data.id].reject(data.data);
                    break;
                default:
            }
        });
    }
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send("start", this.config);
        });
    }
    lift() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send("lift", this.config);
        });
    }
    send(command, data) {
        return __awaiter(this, void 0, void 0, function* () {
            const id = this.getMessageId();
            const promise = new Promise((resolve, reject) => {
                this.commands[id] = {
                    resolve,
                    reject,
                };
            });
            this.child.send({
                id,
                command,
                data,
            });
            return promise;
        });
    }
    getMessageId() {
        return (this.id++).toString();
    }
    kill() {
        return __awaiter(this, void 0, void 0, function* () {
            this.child.kill();
            return Promise.resolve();
        });
    }
    create(modelId, created) {
        return this.send("create", {
            modelId,
            created,
        });
    }
    update(modelId, params, updated) {
        return this.send("update", {
            modelId,
            params,
            updated,
        });
    }
}
exports.RemoteApp = RemoteApp;
;
