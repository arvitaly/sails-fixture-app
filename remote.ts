import { ChildProcess, fork } from "child_process";
import { IStartConfig } from "./typings";
interface ICommand {
    resolve: (data?: any) => any;
    reject: (err?: any) => any;
}
export async function start(config: IStartConfig = {}) {
    const app = new RemoteApp(config);
    await app.start();
    return app;
}
export async function lift(config: IStartConfig = {}) {
    const app = new RemoteApp(config);
    await app.lift();
    return app;
}
export class RemoteApp {
    protected commands: { [index: string]: ICommand } = {};
    protected id = 0;
    protected child: ChildProcess;
    constructor(public config: IStartConfig = {}) {
        this.child = fork(__dirname + "/remote-client");
        this.child.on("message", (data) => {
            switch (data.type) {
                case "log":
                    console.warn(data.data);
                    break;
                case "resolve":
                    this.commands[data.id].resolve(data.data);
                    break;
                case "reject":
                    this.commands[data.id].reject(data.data);
                    break;
                default:
            }
        });
    }
    public async start() {
        return this.send("start", this.config);
    }
    public async lift() {
        return this.send("lift", this.config);
    }
    public async send(command: string, data?: any) {
        const id = this.getMessageId();
        const promise = new Promise<any>((resolve, reject) => {
            this.commands[id] = {
                resolve,
                reject,
            };
        });
        this.child.send({
            id,
            command,
            data,
        });
        return promise;
    }
    public getMessageId() {
        return (this.id++).toString();
    }
    public async kill() {
        this.child.kill();
        return Promise.resolve();
    }
    public create(modelId: string, created: any) {
        return this.send("create", {
            modelId,
            created,
        });
    }
    public update(modelId: string, params: any, updated: any) {
        return this.send("update", {
            modelId,
            params,
            updated,
        });
    }
};
