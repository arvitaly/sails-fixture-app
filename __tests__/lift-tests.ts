import httpRequest = require("request");
import { createModel1 } from "./../";
import { lift, RemoteApp } from "./../remote";
describe("Lift tests", () => {
    let app: RemoteApp;
    beforeEach(async () => {
        app = await lift();
    });
    afterEach(async () => {
        await app.kill();
    });
    it("http create and get model", async () => {
        let result = await postRequest("http://127.0.0.1:14001/modelName1", createModel1());
        result = await getRequest("http://127.0.0.1:14001/modelName1");
        delete result.createdAt;
        delete result.updatedAt;
        expect(result).toMatchSnapshot();
    });
});
function postRequest(url: string, params?: any) {
    const options = {
        method: "POST",
        url,
        headers: {
            "Content-Type": "application/json",
        },
        json: params,
    };
    return new Promise<any>((done, reject) => {
        httpRequest.post(options, (error, response, body) => {
            if (error) {
                reject(error);
                return;
            }
            if (response.statusCode !== 200 && response.statusCode !== 201) {
                reject("Invalid status code: " + response.statusCode);
                return;
            }
            done(body);
        });
    });
}
function getRequest(url: string, params?: any): Promise<any> {
    return new Promise<any>((done, reject) => {
        httpRequest(url, params, (error, response, body) => {
            if (error) {
                reject(error);
                return;
            }
            if (response.statusCode !== 200) {
                reject("Invalid status code: " + response.statusCode);
                return;
            }
            done(body);
        });
    });
};
