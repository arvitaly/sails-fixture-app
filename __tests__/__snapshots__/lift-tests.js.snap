exports[`Lift tests http create and get model 1`] = `"[{\"model3s\":[],\"firstActive\":\"2016-11-18T11:25:11.000Z\",\"id\":1,\"isActive\":false,\"name\":\"na1\",\"num\":15,\"updatedAt\":\"2016-11-18T11:25:11.000Z\",\"createdAt\":\"2016-11-18T11:25:11.000Z\"}]"`;
